### ISSUE TITLE
StoragePath in the context of Multiroot

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/vscode#37137]
The `ExtensionContext` object passed to the activate function provides access to a `storgePath` which extensions can use to store workspace local data in the file system using Node's FS API.

In a multi root setup the path contains a unique path to a location. However extensions would like to read and write the meta data for a workspace folder to share the information across workspace and a folder workspace. 

We should therefore offer an API `getStoragePath` with takes a workspace folder and returns the `storagePath` for that workspace folder.

The interesting question here is whether we need to do the same for workspaceState. IMO it would make sense that a extension wants to share state between the same location being opened as a workspace folder across different workspaces and a folder workspace.

@bpasero any comments
@jrieken regarding the API proposal getStoragePath.

**Comment by jrieken** [comment]
> to read and write the meta data for a workspace folder to share the information across workspace and a folder workspace.

Not sure that sharing `storagePath` between folders is the right thing to do. The idea is still that we cleanup storage when a folder doesn't exist anymore. By sharing this will become more complex and we are likely breaking workspace B by removing/cleaning workspace A. Extensions that want to share (disk) state between workspace folders could  create their "own" folder in their "own" location, store that location globally, and access it in their workspaces. This issue reminds me of #22557 for which I have suggested something similar

**Comment by bpasero** [comment]
Another thing to worth thinking about is that as soon as multiple extension hosts access the same storagePath you will potentially have multiple processes fighting over the same resources. Today you will never have the situation that the same extension host accesses the same storage path (or local storage) more than once because you cannot open the same folder in multiple windows.

**Comment by dbaeumer** [comment]
@jrieken: I think I didn't express myself correctly. It is not about sharing storage between workspace A and workspace B. It is about that the storage of a workspace folder F1 can be seen in a workspace A and a workspace B if both have the workspace folder F1 added. 

I agree about the 'locking' problem if the workspace folder F1 is opened in workspace W1 and workspace W2. But that problem also exists if we ask them to manage the storage by themselves.

I agree that the problem goes away if we give every workspace its own storage but it makes it a lot harder to clean that storage. We could never and would start to rely on the extension to clean the storage up. Consider the following sequence:

- workspace W
- add folder f1
- add folder f2
- an extension E has generated state for the folder f1 and f2
- remove folder f1

The only hint that the storage for f1 in workspace W is not needed anymore can come from Extension E since we wouldn't enforce any storage layout underneath storage path for W/E

I know that we have a general problem with cleaning the storage but so far we can at least detect the cases where a storage path is necessary (since the workspace / workspace folder exists). However we need user interaction for storage that might be obsolete (due to file system mounting).

Another cons for having the storage with the workspace only is sharing / time to compute it. If the data is huge and expensive to compute it is duplicated for every workspace that contains that folder.

**Comment by bpasero** [comment]
I don't think the problem of multiple processes fighting over the same resources exist if we keep the storagePath separate for each single-folder workspace and each multi-root-workspace because also today you can not open the same workspace in multiple windows.

If all state is associated with a workspace folder, could this lead to problems where specific state is only valid in certain workspaces but not in all? E.g. what if there is state that is only there because of the presence of another workspace folder. Wouldn't it be hard for an extension to do the right thing then if all state is shared?

Maybe we need to introduce a new thing besides local storage and storage path that is just scoped to the extension id and not to any workspace. An extension is then free to put files into that folder in any way desired and can do sharing between workspace folders. It would be up to the extension to build this in a way that it gets cleaned up from time to time and works in a multi-process environment too.

**Comment by jrieken** [comment]
> If the data is huge and expensive to compute it is duplicated for every workspace that contains that folder.

If those things are given then workspace-storage is probably not the right place to begin with...

> Maybe we need to introduce a new thing besides local storage and storage path that is just scoped to the extension id and not to any workspace. 

Sounds like `globalStoragePath` in addition to `storagePath` which maybe should have been `workspaceStoragePath`... Tho the clean-up might not be simple. We cleanup when you remove/uninstall an extension but other cases we need the extension to do something

**Comment by bpasero** [comment]
Yeah I like the idea of `globalStoragePath` vs. `workspaceStoragePath`

**Comment by dbaeumer** [comment]
I think conceptually there is no difference between a `globalStoragePath` and a `workspaceStoragePath` in terms of us picking a location.

Our settings are basically the same. They are  scoped by user / global and resource / folder. Lucky us we store the resource settings with the workspace folder so we have no problems to clean them up. They disappear when the user deletes the folder :-). We introduced the storagePath since we didn't want extensions to write into the .vscode folder (which may be we shouldn't have done either).

Things I know: if we provide storage locations we should structure the storage so that we can help cleaning it up when a folder gets removed from a workspace or a workspace folder gets deleted from disk or a extensions gets uninstalled.

Things I am unsure about: if the storage of a workspace folder should be sharable across workspaces. We do share the settings so it sounds logical to share the storage as well.  However I see all the problems that arise from concurrent access to that storage. The problem I see from not sharing the storage is that the content of the storage mostly depends on settings. So extensions need to remember for which settings the storage got computed and re-compute the storage if such settings change.

**Comment by bpasero** [comment]
Btw on top of `storagePath` there is also users asking for a way to access a specific `workspaceState` memento of a specific workspace folder. To me, whatever we do for `storagePath` should also apply to the UI-state an extension can read and write.

I like the idea of keeping this symmetric with our settings where you can access user, workspace and workspace folder levels. It would be nice if these scopes could be used for both storage path and memento state. We would just have to make sure to store these things properly on disk somewhere.

The challenges of concurrent access and cleanup remain, no matter what, unless we would actually store this data within the workspace folder (like settings).

**Comment by alefragnani** [comment]
I talked to @bpasero about that. In my case, I would like to have access to `workspaceState` for specific folders because when a folder is _added_ to some workspace, its _state_ is not migrated/imported. If the same _migration_ occured for any folder _added_ to workspace, it would work as well.

Thanks in advance
